packages/fuel-gauge/src/transaction-response.test.ts

```diff
@@ -9,6 +9,75 @@ import {
   randomBytes,
   WalletUnlocked,
 } from 'fuels';
+import type { MockInstance } from 'vitest';
+
+async function verifyKeepAliveMessageWasSent(subscriptionStream: ReadableStream<Uint8Array>) {
+  const decoder = new TextDecoder();
+  const reader = subscriptionStream.getReader();
+  let hasKeepAliveMessage = false;
+  do {
+    const { value, done } = await reader.read();
+    if (done) {
+      break;
+    }
+    const text = decoder.decode(value);
+    if (text === ':keep-alive-text\n\n') {
+      hasKeepAliveMessage = true;
+    }
+  } while (!hasKeepAliveMessage);
+
+  // The keep-alive message is sent every 15 seconds,
+  // and this assertion verifies that it was indeed sent.
+  // if this fails, check if the duration was changed on the fuel-core side.
+  // As of the time of writing, the latest permalink where this info can be found is:
+  // https://github.com/FuelLabs/fuel-core/blob/bf1b22f47c58a9d078676c5756c942d839f38916/crates/fuel-core/src/graphql_api/service.rs#L247
+  // To get the actual latest info you need to check out the master branch:
+  // https://github.com/FuelLabs/fuel-core/blob/master/crates/fuel-core/src/graphql_api/service.rs#L247
+  // This link can fail because master can change.
+  expect(hasKeepAliveMessage).toBe(true);
+}
+
+function getSubscriptionStreamFromFetch(streamHolder: { stream: ReadableStream<Uint8Array> }) {
+  function getFetchMock(
+    fetchSpy: MockInstance<
+      [input: RequestInfo | URL, init?: RequestInit | undefined],
+      Promise<Response>
+    >
+  ) {
+    return async (...args: Parameters<typeof fetch>) => {
+      /**
+       * We need to restore the original fetch implementation so that fetching is possible
+       * We then get the response and mock the fetch implementation again
+       * So that the mock can be used for the next fetch call
+       */
+      fetchSpy.mockRestore();
+      const r = await fetch(...args);
+      fetchSpy.mockImplementation(getFetchMock(fetchSpy));
+
+      const isSubscriptionCall = args[0].toString().endsWith('graphql-sub');
+      if (!isSubscriptionCall) {
+        return r;
+      }
+
+      /**
+       * Tee duplicates a stream and all writes happen to both streams.
+       * We can thus use one stream to verify the keep-alive message was sent
+       * and pass the other forward in place of the original stream,
+       * thereby not affecting the response at all.
+       * */
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const [stream1, stream2] = r.body!.tee();
+      // eslint-disable-next-line no-param-reassign
+      streamHolder.stream = stream1;
+      return new Response(stream2);
+    };
+  }
+
+  const fetchSpy = vi.spyOn(global, 'fetch');
+
+  fetchSpy.mockImplementation(getFetchMock(fetchSpy));
```

Comment:
Are we using mocks inside `fuel-gauge` now? 🤔

Reply:
Ah, this is the abstract part :D I agree, it's weird, to say the least. I couldn't figure out a nicer way to _not_ change the behavior while still getting what I need :smile:.

The purpose of this code is to protect the test from false positives if the node's default 15 seconds for sending keep-alive messages was increased over the test's `--poa-interval-period 17s` and thus no keep-alive message was sent. We'd be thinking the test covers keep-alive messages, but it wouldn't be.

As for the implementation and why I "mocked", I couldn't read keep-alive messages any other way because they are ignored by the code and the streams get closed after they are read so I couldn't go with just a simple spy.

In order to get the underlying stream that contains the subscription responses before it gets closed (actual responses and the keep-alive message) and verify that a keep-alive message was sent, I had to "mock" fetch. The mock itself isn't changing anything, it's just duplicating the stream in the case of a subscription so that I can inspect it for the keep-alive message:
- If the call is not to a subscription endpoint, the response will be just as a regular `fetch` response would be. 
- If the call is to a subscription endpoint, the response will _still_ be the same as that of a regular `fetch`, but the stream is duplicated so that I can parse it independently in `verifyKeepAliveMessageWasSent`.

 The `getFetchMock` function looks weird with its `fetchSpy.mockRestore()` and `fetchSpy.mockImplementation(getFetchMock(fetchSpy))` calls precisely because I didn't want to change `fetch` behavior. I had to make sure that the first and _every_ subsequent fetch call are "mocked" with the logic described above so that, for the specific subscription fetch call, I can duplicate the stream before it's closed.